ovos_utils/log.py

```diff
@@ -37,65 +29,97 @@ class LOG:
-    info = _make_log_method(logging.Logger.info)
-    warning = _make_log_method(logging.Logger.warning)
-    error = _make_log_method(logging.Logger.error)
-    exception = _make_log_method(logging.Logger.exception)
+    base_path = "stdout"
+    fmt = '%(asctime)s.%(msecs)03d - ' \
+          '%(name)s - %(levelname)s - %(message)s'
+    datefmt = '%Y-%m-%d %H:%M:%S'
+    formatter = logging.Formatter(fmt, datefmt)
+    max_bytes = 50000000
+    backup_count = 3
+    name = 'OVOS'
+    level = "DEBUG"
+    _loggers = {}
 
     @classmethod
-    def init(cls):
-        cls.level = "DEBUG"
-        fmt = '%(asctime)s.%(msecs)03d - ' \
-              '%(name)s - %(levelname)s - %(message)s'
-        datefmt = '%H:%M:%S'
-        formatter = logging.Formatter(fmt, datefmt)
-        cls.handler = logging.StreamHandler(sys.stdout)
-        cls.handler.setFormatter(formatter)
-
-        # Enable logging in external modules
-        cls.create_logger('').setLevel(cls.level)
+    def init(cls, config=None):
+        config = config or {}
+        cls.base_path = config.get("path", "stdout")
+        cls.max_bytes = config.get("max_bytes", 50000000)
+        cls.backup_count = config.get("backup_count", 3)
+        cls.level = config.get("level", "DEBUG")
 
     @classmethod
-    def create_logger(cls, name):
+    def create_logger(cls, name, tostdout=True):
+        if name in cls._loggers:
+            return cls._loggers[name]
         logger = logging.getLogger(name)
         logger.propagate = False
-        logger.addHandler(cls.handler)
+        # always log to stdout
+        if tostdout:
+            stdout_handler = logging.StreamHandler(sys.stdout)
+            stdout_handler.setFormatter(cls.formatter)
+            logger.addHandler(stdout_handler)
+        # log to file
+        if cls.base_path != "stdout":
+            path = join(cls.base_path,
+                        cls.name.lower().strip() + ".log")
+            handler = RotatingFileHandler(path, maxBytes=cls.max_bytes,
+                                          backupCount=cls.backup_count)
+            handler.setFormatter(cls.formatter)
+            logger.addHandler(handler)
+        logger.setLevel(cls.level)
+        cls._loggers[name] = logger
         return logger
 
-    def __init__(self, name):
-        LOG._custom_name = name
+    @classmethod
+    def set_level(cls, level):
+        cls.level = level
+        for l in cls._loggers:
+            cls._loggers[l].setLevel(level)
+
+    @classmethod
+    def _get_real_logger(cls):
+        name = ""
+        if cls.name is not None:
+            name = cls.name + " - "
+
+        # Stack:
+        # [0] - _log()
+        # [1] - debug(), info(), warning(), or error()
+        # [2] - caller
+        stack = inspect.stack()
+
+        # Record:
+        # [0] - frame object
+        # [1] - filename
+        # [2] - line number
+        # [3] - function
+        # ...
+        record = stack[2]
```

Comment:
Still don't get it. I've never been able to get `inspect.stack()` to do anything remotely useful, nor have I discerned any reliability to the order in which things appear on the stack.

Reply:
I mean, if it works consistently in Mycroft, and it doesn't crash, I guess it is what it is, but counting on the stack to have exactly what you expect on top seems like *madness*, I don't understand how it's failed to fail for 3+ years

Reply:
Maybe if this were C++...

Reply:
this is immediately the line before this func call, so its consistent. And we extract the module/line where log is happening

i trust @forslund on this one :)